detection/opencv_detector.go

```diff
@@ -0,0 +1,400 @@
+func (d *OpenCVDetector) tryLoadModel(weightsPath, configPath, modelName string) bool {
+	// Check if both files exist
+	if _, err := os.Stat(weightsPath); os.IsNotExist(err) {
+		return false
+	}
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
+		return false
+	}
+
+	// Try to load the network
+	net := gocv.ReadNetFromDarknet(configPath, weightsPath)
+	if net.Empty() {
+		log.Printf("Failed to load %s model from %s", modelName, weightsPath)
+		return false
+	}
+
+	// Set backend and target for optimal performance on Raspberry Pi
+	net.SetPreferableBackend(gocv.NetBackendOpenCV)
+	net.SetPreferableTarget(gocv.NetTargetCPU)
+
+	d.net = &net
+	log.Printf("Successfully loaded %s model for object detection", modelName)
+	return true
+}
+
+// loadClassNames loads COCO class names from file
+func (d *OpenCVDetector) loadClassNames(classesPath string) error {
+	data, err := os.ReadFile(classesPath)
+	if err != nil {
+		return err
+	}
+
+	classes := strings.Split(strings.TrimSpace(string(data)), "\n")
+	d.classes = classes
+	return nil
+}
+
+// DetectObjects performs object detection on an image file
+func (d *OpenCVDetector) DetectObjects(imagePath string) ([]DetectionBox, error) {
+	if !d.initialized || d.net == nil {
+		return nil, fmt.Errorf("OpenCV detector not properly initialized")
+	}
+
+	// Load the image in color mode (BGR) - YOLO models require 3-channel color input
+	// IMReadColor ensures we get BGR color data rather than grayscale, which is essential for accurate object detection
+	img := gocv.IMRead(imagePath, gocv.IMReadColor)
+	if img.Empty() {
+		return nil, fmt.Errorf("failed to load image: %s", imagePath)
+	}
+	defer img.Close()
+
+	return d.detectObjectsFromMat(img)
+}
+
+// detectObjectsFromMat performs detection on a gocv.Mat
+func (d *OpenCVDetector) detectObjectsFromMat(img gocv.Mat) ([]DetectionBox, error) {
+	// Create blob from image - converts image to YOLO input format
+	// Normalizes pixel values (1.0/255.0), resizes to 416x416, and arranges data layout for neural network processing
+	blob := gocv.BlobFromImage(img, 1.0/255.0, image.Pt(416, 416), gocv.NewScalar(0, 0, 0, 0), true, false, gocv.MatTypeCV32F)
+	defer blob.Close()
+
+	// Set input to the network
+	d.net.SetInput(blob, "")
+
+	// Get output layer names
+	outputNames := d.net.GetUnconnectedOutLayersNames()
+
+	// Run forward pass
+	outputs := d.net.ForwardLayers(outputNames)
+	defer func() {
+		for i := range outputs {
+			outputs[i].Close()
+		}
+	}()
+
+	// Process detections
+	detections := d.processDetections(outputs, img.Cols(), img.Rows())
+
+	// Apply Non-Maximum Suppression (NMS)
+	filteredDetections := d.applyNMS(detections, 0.5, 0.4)
+
+	return filteredDetections, nil
+}
+
+// processDetections extracts detection boxes from network outputs
+func (d *OpenCVDetector) processDetections(outputs []gocv.Mat, imgWidth, imgHeight int) []DetectionBox {
+	var detections []DetectionBox
+	confidenceThreshold := float32(0.5)
```

Comment:
move this into a function parameter. callers should decide what the confidence threshold is

Reply:
let the caller pass a default confidence threshold of 60%